KAS-3307: Combine {Edit,New}MeetingModal into one component

sergiofenoll opened the pull request:
Combines the two modal components into a single component which takes a meeting record and enables a user to change its properties/relationships as was previously possible in the `NewMeetingModal` component.

The caller is responsible for passing in either a "real" meeting record, i.e. of a meeting that has been fetched from the backend, or a "new" meeting record, i.e. an object instantiated with `store.createRecord('meeting')`. The component makes use of the record's `isNew` member to determine whether it should perform the extra steps for a meeting creation when saving.

In theory this behaviour could be extracted to the caller as well (i.e., the `agendas.overview` controller creates the agendaitem for the meeting once it's been created), leaving this component purely as a meeting editor. Maybe that's a cleaner solution.

Other changes in this PR:

- The modal changes can't be saved unless all necessary fields are filled in.
- The `numberRepresentation` input field can be emptied (but not saved in this sate)
- The `meetingNumber` input field can be emptied (but not saved in this state)

The last two are to prevent behaviour where emptying the input caused it to be reset to the default value, which looks weird.

sergiofenoll committed:
Combine {Edit,New}MeetingModal into one component

sergiofenoll committed:
Disable saving if meetingNumber or meetingRepresenation are not present
for all meeting types

sergiofenoll committed:
Remove unused required

`required` attribute will only work for forms, which we don't use

sergiofenoll committed:
Merge branch 'development' into feature/KAS-3307-type-van-vergadering-correct-kunnen-wijzigen-naar-pvv

sergiofenoll committed:
Use numberRepresentation, not formattedMeetingIdentifier

sergiofenoll committed:
Reorder initializing of numberRepresentation and meetingNumber

sergiofenoll committed:
Use correct chain in Cypress

A meeting's numberRepresentation is no longer an input field by default,
so we need to use `should('eq', value)` instead of `should('have.value',
value)`.

sergiofenoll committed:
Use `should('contain')` instead of `should('eq')`

erikap commented on app/components/meeting/edit-meeting-modal.hbs:
```diff
@@ -1,61 +1,130 @@
-  <Auk::Modal::Header @title={{t "edit-meeting"}} @onClose={{@onCancel}} />
+  <Auk::Modal::Header
+    @title={{if @meeting.isNew (t "new-agenda") (t "edit-meeting")}}
+    @closeAble={{true}}
```
`closeable` instead of `closeAble`? But since `true` is the default it can be omitted.

erikap commented on app/components/meeting/edit-meeting-modal.js:
```diff
@@ -2,39 +2,60 @@ import Component from '@glimmer/component';
+
+  @tracked selectedMainMeeting;
+  @tracked isFinal;
   @tracked kind;
```
Rename to `selectedKind` to be in line with `selectedMainMeeting`?

erikap commented on app/components/meeting/edit-meeting-modal.js:
```diff
@@ -47,40 +68,175 @@ export default class MeetingEditMeetingComponent extends Component {
 
   @action
-  setKind(kind) {
+  async setKind(kind) {
```
Rename to `selectKind` to be in line with `selectMainMeeting`?

erikap commented on app/pods/agendas/controller.js:
```diff
@@ -20,11 +22,14 @@ export default class AgendasController extends Controller.extend(DefaultQueryPar
   @action
   successfullyAdded() {
     this.isCreatingNewSession = false;
+    this.newMeeting = this.store.createRecord('meeting');
```
Maybe better to move to `setupController` in the route since we want to reinitialize this each time the user visits the route?

ValenberghsSven replied:
See other comment, do we need to initialize this when visiting route?
We could just create a model for meeting before opening the `editMeetingModal`.

erikap commented on app/pods/agendas/controller.js:
```diff
@@ -7,10 +7,12 @@ import { action } from '@ember/object';
   @service router;
 
   @tracked isCreatingNewSession = false;
+  @tracked newMeeting = this.store.createRecord('meeting');
```
Maybe better to move to `setupController` in the route since we want to reinitialize this each time the user visits the route?

erikap commented on app/components/meeting/edit-meeting-modal.hbs:
```diff
@@ -1,61 +1,130 @@
+          <WebComponents::VlDatepicker
+            @enableTime={{true}}
+            @date={{array this.startDate}}
+            @dateChanged={{this.selectStartDate}}
```
Using inline `set` like [here](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/18b3af8f59030c17dcc3d117c4aa3fa344ff3009/app/components/publications/new-publication-modal.hbs#L109)

erikap commented on app/components/meeting/edit-meeting-modal.js:
```diff
@@ -47,40 +68,175 @@ export default class MeetingEditMeetingComponent extends Component {
+  @task
+    *initializeMainMeeting() {
+      this.selectedMainMeeting = yield this.args.meeting.mainMeeting;
+    }
```
Indentation

erikap commented on app/components/meeting/edit-meeting-modal.js:
```diff
@@ -47,40 +68,175 @@ export default class MeetingEditMeetingComponent extends Component {
   @task
   *initializeKind() {
     this.kind = yield this.args.meeting.kind;
+    this.kind ??= yield this.store.findRecordByUri('concept', CONSTANTS.MEETING_KINDS.MINISTERRAAD);
```
Didn't know this `??=` syntax. This is the same as `this.kind = (yield this.args.meeting.kind) || (yield this.store.findRecordByUri('concept', CONSTANTS.MEETING_KINDS.MINISTERRAAD);`, right? (just to be sure I understood it correctly)

sergiofenoll replied:
yes, or more specifically `(yield this.args.meeting.kind) ?? (yield this.store...` since it only checks for null/undefined instead of all false-y values. but writing it like that results in `this.kind` being `null` when creating a new meeting. I suspect it's because the evaluation of `||` (or `??`) happens against the promises instead of the values themselves. this results in the first promise being accepted as non-false-y (or non-nullish), but once it resolves it turns out to be `null` after all. that's why i split it up in two statements, so that the promises are settled as expected.

erikap commented on app/components/meeting/edit-meeting-modal.js:
```diff
@@ -2,39 +2,60 @@ import Component from '@glimmer/component';
+    this.startDate = this.args.meeting.plannedStart || new Date(now.getFullYear(), now.getMonth(), now.getDate(), 10, 0, 0);
     this.extraInfo = this.args.meeting.extraInfo;
-    this.meetingNumber = this.args.meeting.number;
+    this.isFinal = this.args.isFinal ?? false;
```
`this.isFinal` seems not to be used. Argument `isFinal` is never passed to the component.

erikap replied:
Still confused about the `this.isFinal`. Since `this.args.isFinal` is always `undefined` (because not passed to the component) `this.isFinal` will always be `false`. The property is nowhere manipulated; only used to set `this.args.meeting.isFinal = this.isFinal` when saving the meeting. Is the tracked property needed? And does it need to be set to `false` on each save of the meeting? I don't see how it relates to the saved changes.

Behaviour I would expect is `isFinal` to be `false` on creation of a new meeting (that can be covered by adding the property on `createRecord('meeting', ...)` and leaving it untouched on edit of the meeting? Or am I missing something here @ValenberghsSven ?

sergiofenoll replied:
That line should be `this.isFinal = this.args.meeting.isFinal ?? false`, so it follows the expected behaviour, I just missed the `.meeting`

erikap commented on app/components/meeting/edit-meeting-modal.js:
```diff
@@ -47,40 +68,175 @@ export default class MeetingEditMeetingComponent extends Component {
     this.args.meeting.number = this.meetingNumber;
     this.args.meeting.numberRepresentation = this.numberRepresentation;
+    this.args.meeting.mainMeeting = this.selectedMainMeeting;
+    this.args.meeting.isFinal = false;
```
A bit weird to overwrite this flag while editing the meeting. Was it like this before?

sergiofenoll replied:
was meant to use `this.isFinal` here, thanks for catching that!

erikap reviewed (changes requested):
Since behaviour greatly differs for new/existing meeting, I have a preference for moving the creation of the agenda/first agenda item to the `agendas.overview` controller as you suggest.

sergiofenoll committed:
Refactor edit-meeting-modal

Extract extra steps related to new meeting creation to agendas
controller. Rename `kind` -> `selectedKind` to be like
`selectedMainMeeting`. Extract new meeting record creation to
`setupController` hook of the route.

ValenberghsSven commented:
1 of the failing tests seems legit in decision spec. We are creating PVV agenda right after the main agenda.
I'm assuming we end up editing the first agenda to be PVV rather then creating a new PVV agenda.
need to confirm manually.

sergiofenoll commented:
> 1 of the failing tests seems legit in decision spec. We are creating PVV agenda right after the main agenda. I'm assuming we end up editing the first agenda to be PVV rather then creating a new PVV agenda. need to confirm manually.

Yup, you're right. Currently checking it out. It's because after saving the meeting, the `setupController` hook doesn't get called again, so when we create another meeting we're in fact just editing the previous meeting.

In checking this, I also noticed that the title of the modal changes from "Nieuwe agenda aanmaken" to "Vergadering wijzigen", because the record gets saved. Will make the check on the title more explicit as well, with an `isNew` argument

sergiofenoll committed:
Create new empty meeting record after saving

Also explicitly tell component if the meeting is new to prevent
mid-saving changes to the modal title

erikap commented on app/pods/agendas/controller.js:
```diff
@@ -20,12 +27,86 @@ export default class AgendasController extends Controller.extend(DefaultQueryPar
+    }
+    await this.newsletterService.createNewsItemForMeeting(this.newMeeting);
+    this.newMeeting = this.store.createRecord('meeting');
+    // TODO: Should fix sessionNrBug
+    // Import from meeting-utils.js
+    // await assignNewSessionNumbers();
```
Is this comment still applicable @ValenberghsSven (and if yes, what does it entail) or can it be removed?

ValenberghsSven replied:
Not sure what that bug is or was.
It seems that the relevant code in the session-number-service would delete the triple and reinsert a decimal `adms:identifier` (`meeting.number` in frontend). Knowing what we know now, decimal is not a correct type since we should use integer. 
So the comment can go, but we might have to look into cleaning up the type differences (and maybe find the bug) in new ticket?

erikap commented on app/pods/agendas/controller.js:
```diff
@@ -20,12 +27,86 @@ export default class AgendasController extends Controller.extend(DefaultQueryPar
+    const agenda = await this.createAgenda(this.newMeeting);
+
+    const closestMeeting = await fetchClosestMeetingAndAgendaId(this.newMeeting.plannedStart);
+    const broader = await this.newMeeting.kind?.get('broader');
```
Try to avoid usage of `get()` but explicitly await promises (even if you now it will immediately resolve because the data has already been fetched)
```javascript
const kind = await this.newMeeting.kind;
const broader = await kind?.broader;
```

erikap commented on app/pods/agendas/controller.js:
```diff
@@ -20,12 +27,86 @@ export default class AgendasController extends Controller.extend(DefaultQueryPar
 
   @action
-  successfullyAdded() {
+  async successfullyAdded() {
```
`createAgendaAndNewsletter` is maybe a better name to cover the content?

erikap commented on app/pods/agendas/template.hbs:
```diff
@@ -26,7 +26,9 @@
 {{#if this.isCreatingNewSession}}
-  <Meeting::NewMeetingModal
+  <Meeting::EditMeetingModal
+    @isNew={{true}}
```
Maybe better to embed this logic in `Meeting::EditMeetingModal` by adding a tracked property `isNew` in the meeting modal component which is initialized in its constructor based on `this.args.meeting.isNew`. Keeps the component interface simpler.

sergiofenoll committed:
EditMeetingModal interface improvements

Rename meeting save action in route & track isNew in EditMeetingModal
component instead of in agendas route

sergiofenoll committed:
Merge branch 'development' into feature/KAS-3307-type-van-vergadering-correct-kunnen-wijzigen-naar-pvv

ValenberghsSven commented on app/components/meeting/edit-meeting-modal.js:
```diff
@@ -47,40 +67,99 @@ export default class MeetingEditMeetingComponent extends Component {
     this.args.meeting.number = this.meetingNumber;
     this.args.meeting.numberRepresentation = this.numberRepresentation;
+    this.args.meeting.mainMeeting = this.selectedMainMeeting;
+    this.args.meeting.isFinal = this.isFinal;
```
The usage of "isFinal" in this component is mainly for applying the default of `isFinal:false` to a new meeting I guess?
Could we apply the defaults to the creation of the model from args instead?

Just want to prevent us accidentally overwriting the isFinal value + you can't actually change this value in this component

ValenberghsSven commented on app/pods/agendas/controller.js:
```diff
@@ -20,12 +27,87 @@ export default class AgendasController extends Controller.extend(DefaultQueryPar
   closeNewSessionModal() {
     this.isCreatingNewSession = false;
+    this.newMeeting.deleteRecord();
+    this.newMeeting = this.store.createRecord('meeting');
```
To me, it doesn't really make sense to create something on closing of a modal.
also, doing it this way is resulting in us having 3 places where we create a meeting (here, in `createAgendaAndNewsletter` and in the route's `setupController`)
Could we just create the meeting (with defaults, see other comments) on the opening of the modal `openNewSessionModal`? Should be the only place

ValenberghsSven commented on app/components/meeting/edit-meeting-modal.js:
```diff
@@ -2,39 +2,59 @@ import Component from '@glimmer/component';
+
   constructor() {
     super(...arguments);
+    this.isNew = this.args.meeting.isNew;
```
Instead of relying on the ember model to be new, we could pass `@isNew={{true}}` when creating a new one. We just have to default to false for editing

sergiofenoll replied:
@erikap [asked for the opposite change](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1296#discussion_r856082386) (i.e. don't pass `isNew`, compute it in the component)

I see the value in both, but don't have a strong preference for one or the other.
Ik zou zeggen, vecht het tussen elkaar uit :stuck_out_tongue_winking_eye:

ValenberghsSven replied:
goh, ik maak niet zo vaak gebruik van de ember model properties buiten `hasDirtyAttributes`.
Ik bekijk het eerder als volgt:
Bij creatie van een meeting is dit true.
Bij edit van een meeting is dit false.
In dat opzicht zullen opties beiden werken. Voor mij is het gelijk.
Voor de ember properties moet je weten dat ze bestaan en wat ze juist doen om ze correct te gebruiken.

ValenberghsSven replied:
Hou het maar zoals het is.

ValenberghsSven reviewed (changes requested):
Template `new-meeting-modal` still exists, guess we can delete?

sergiofenoll committed:
Remove NewMeetingModal template

ValenberghsSven committed:
Merge branch 'development' of github.com:kanselarij-vlaanderen/kaleidos-frontend into feature/KAS-3307-type-van-vergadering-correct-kunnen-wijzigen-naar-pvv

sergiofenoll committed:
Create meeting record in one place & pass isFinal

sergiofenoll committed:
Merge branch 'feature/KAS-3307-type-van-vergadering-correct-kunnen-wijzigen-naar-pvv' of github.com:kanselarij-vlaanderen/frontend-kaleidos into feature/KAS-3307-type-van-vergadering-correct-kunnen-wijzigen-naar-pvv

sergiofenoll committed:
Merge branch 'development' into feature/KAS-3307-type-van-vergadering-correct-kunnen-wijzigen-naar-pvv
